File: c/linked_list.c
#include <stdio.h>

typedef struct Node
{
    char *value;
    struct Node *next;
} Node;

typedef struct LinkedList
{
    Node *head;
    Node *tail;
    unsigned int length;
} LinkedList;

Node *get_node(LinkedList *lst, int index)
{
    if (index >= lst->length || index < 0)
        return NULL;
    int i = 0;
    Node *current = lst->head;
    while (i != index)
    {
        i++;
        current = current->next;
    }
    return current;
}

void print_linked_list(LinkedList *lst)
{
    Node *ptr = lst->head;
    while (ptr->next != NULL)
    {
        printf("%s -> ", ptr->value);
        ptr = ptr->next;
    }
    printf("%s -> NULL\n", ptr->value);
}

void append_node(LinkedList *lst, Node *node)
{
    if (lst->head == NULL)
    {
        lst->head = node;
        lst->tail = lst->head;
    }
    else
    {
        lst->tail->next = node;
        lst->tail = node;
    }
    lst->length++;
}

void insert(LinkedList *lst, Node *node, int index)
{
    int i = 0;
    if (index == 0)
    {
        node->next = lst->head;
        lst->head = node;
        lst->length++;
        return ;
    }
    else if (index >= lst->length)
    {
        append_node(lst, node);
        return ;
    }    
    Node *ptr = lst->head;
    while (i < index - 1)
    {
        ptr = ptr->next;
        i++;
    }
    node->next = ptr->next;
    ptr->next = node;
    lst->length++;
}

Node *delete_node(LinkedList *lst, int index)
{
    if (lst->head == NULL)
        return NULL;
    else if (index >= lst->length || index < 0)
        return NULL;
    else if (index == 0)
    {
        Node *result = lst->head;
        lst->head = lst->head->next;
        lst->length--;
        return result;
    }
    int i = 1;
    Node *prev = lst->head;
    Node *curr = lst->head->next;
    while (i < index-1)
    {
        i++;
        prev = curr;
        curr = curr->next;
    }
    if (i == lst->length - 2)
        lst->tail = prev;
    prev->next = curr->next;
    lst->length--;
    return curr;
}

void reverse_list(LinkedList *lst)
{
    if (lst->length == 1)
        return ;
    Node *first = lst->head;
    lst->tail = lst->head;
    Node *second = first->next;
    Node *tmp;
    while (second)
    {
        tmp = second->next;
        second->next = first;
        first = second;
        second = tmp;
    }
    lst->head->next = NULL;
    lst->head = first;
}

int main(void)
{
    LinkedList lst = { NULL, NULL, 1};
    Node a = {"A", NULL};
    append_node(&lst, &a);
    Node b = {"B", NULL};
    Node c = {"C", NULL};
    Node d = {"D", NULL};
    Node e = {"E", NULL};
    Node f = {"F", NULL};
    append_node(&lst, &b);
    printf("Top is  %s\n", lst.tail->value);
    print_linked_list(&lst);
    printf("Head is %s, tail is %s\n", lst.head->value, lst.tail->value);
    append_node(&lst, &c);
    printf("Top is  %s\n", lst.tail->value);
    print_linked_list(&lst);
    printf("Head is %s, tail is %s\n", lst.head->value, lst.tail->value);
    append_node(&lst, &d);
    printf("Top is  %s\n", lst.tail->value);
    print_linked_list(&lst); 
    printf("Head is %s, tail is %s\n", lst.head->value, lst.tail->value);
    append_node(&lst, &e);
    printf("Top is  %s\n", lst.tail->value);
    print_linked_list(&lst);
    printf("Head is %s, tail is %s\n", lst.head->value, lst.tail->value);
    append_node(&lst, &f);
    printf("Top is  %s\n", lst.tail->value);
    print_linked_list(&lst);
    printf("Head is %s, tail is %s\n", lst.head->value, lst.tail->value);
    printf("Deliting..\n");
    delete_node(&lst, lst.length-1);
    print_linked_list(&lst);
    delete_node(&lst, lst.length-1);
    print_linked_list(&lst);
    insert(&lst, &f, 2);
    print_linked_list(&lst);
    return 0;
}

File: python/linked_list.py


class Node:
    def __init__(self, value=0):
        self.data = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def get_item(self, index):
        if index >= self.length:
            return None
        item = self.head
        i = 0
        while (i < index):
            item = item.next
            i += 1
        return item.data

    def print_list(self):
        if self.length == 0:
            print("None")
            return
        head = self.head 
        print(f"{head.data} -> ", end='')
        next = head.next
        while (next != None):
            print(f"{next.data} -> ", end='')
            next = next.next
        print("None")

    def append(self, node: Node):
        if self.head == None:
            self.head = node
            self.tail = self.head
            self.length += 1
        else:
            self.tail.next = node
            self.tail = node
            self.length += 1
        return True
        
    def insert(self, node: Node, index: int) -> None:
        print(f"Inserting...({node.data})")
        if index < 0:
            return False
        if index == 0:
            node.next = self.head
            self.head = node
            self.length += 1
        elif index >= self.length:
            self.append(node)
        else: 
            ptr = self.head
            i = 0
            while (i < index - 1):
                ptr = ptr.next
                i += 1
            node.next = ptr.next
            ptr.next = node
            self.length += 1
        return True
        
    def delete(self, index):
        print("Deliting...")
        if self.head is None:
            return False
        if index >= self.length or index < 0:
            return False
        if index == 0:
            self.head = self.head.next
            self.length -= 1
        else:
            i = 1
            prev = self.head
            curr = self.head.next
            while i < index:
                prev = curr
                curr = curr.next
                i += 1
            if i == self.length - 1:
                self.tail = prev
            prev.next = curr.next
            self.length -= 1
        return True


if __name__ == "__main__":
    linked_list = LinkedList()
    
    a = Node("A")
    b = Node("B")
    c = Node("C")
    d = Node("D")
    
    linked_list.append(a)
    linked_list.print_list()
    linked_list.insert(b, 10)
    linked_list.print_list()
    linked_list.insert(c, 0)
    linked_list.print_list()
    linked_list.insert(d, 0)
    linked_list.print_list()
    linked_list.delete(3)
    linked_list.print_list()
    u = Node("H")
    linked_list.append(u)
    linked_list.print_list()





